lang/LangPrimSource/SC_Levenshstein.h

```diff
@@ -0,0 +1,101 @@
+/*
+    SC_Levenshtein.h
+    Copyright (c) 2019 James Surgenor
+
+    This is a header-only, templated functor to compute the Levenshtein distance between two arrays.
+
+    It uses the reduced memory footprint version, based on Martin Ettl's implementation, and is written as a functor
+    to facilitate even greater memory savings if a single instance is used (although only one comparison can be made
+    at a time). It is self-contained, so could easily be multithreaded.
+
+    Thanks to Brian Heim
+
+    ====================================================================
+
+    SuperCollider real time audio synthesis system
+    Copyright (c) 2002 James McCartney. All rights reserved.
+    http://www.audiosynth.com
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
+*/
+
+#pragma once
+
+#include <vector>
+#include <functional>
+
+template <typename data_type> struct levenshteinDistance {
```

Comment:
typically we use `T` for a generic template parameter name.

i would prefer to see this as a free function -- `matrix` only needs to exist inside the actual computation function, and you're not reusing this object anywhere so you're not saving a memory allocation by keeping it around. you also don't need `min3` as i explain in another comment

Reply:
I was always taught to never use `T`, but happy to change it. Should this be added to the C++ style guide?

In the current use case, we aren't reusing the comparison because of the weak-types.
The benefit comes from other usage where you know you're comparing the same types again and again.
For example, in a parser working with `char`s, we can just have a single distance calculator as a member that won't allocate every time?

Reply:
> I was always taught to never use T, but happy to change it.

interesting, what is the reasoning for that? most C++ documentation (and code for that matter) that i've seen uses CamelCase with an initial capital for template parameters, and a single letter (usually T or N) when the parameter is so generic that there's nothing else to be said about it. the reason for the initial capital is so it's easier to distinguish dependent types from non-dependent ones and from variables whose names generally start with lowercase letters. so, if you write things that way, you have the benefit that other C++ developers are already used to seeing it and writing it, and will immediately know that `T` as a template parameter means "some data type" when reading through a function or class.

> For example, in a parser working with chars, we can just have a single distance calculator as a member that won't allocate every time?

ah, i see your point, i forgot we were considering using this elsewhere! keeping it as a class makes sense to me. could you perhaps leave a brief comment somewhere about that?